### DISCUSSION TITLE
Causal consistency

### DISCUSSION THREAD

**Discussion by em** [vercel/swr#988] [Q&A]
How does SWR guarantee causal consistency (if at all)? I see no mention of this in the docs.

e.g. from the docs:

1. mutate('some/list', {data}, false)  // optimistic 
2. POST some/list {data} // begin request. 
3. wait for response
4.   then mutate('some/list')  // revalidate 

The problem with this workflow is that anything else could cause a separate revalidation between step 2 and 3. That revalidation could finish before the POST does. This would clear the optimistic state, only to have it return again. Your optimistic changes would disappear before the final revalidation finishes, then reappear once it does. And that's assuming the last revalidation finishes last, which itself is not guaranteed. Otherwise your changes would simply disappear entirely until a future revalidation finally gets them.

I fail to see how SWR could possible know that I have a pending POST request and throw away validation responses which turned invalid during the course of being resolved. Since it doesn't even have knowledge of my POST request.

It would need to have an API like this to do that:

mutate({
  key,
  optimistic: synchronous transformer function,
  request: async request function,
  revalidate: true,
})

Am I missing something? Or is this just a design oversight?

**Comment by ErnestBrandi** [comment]
Hi, where is it from the doc ?
From my point of view, `mutate('some/list', {data}, false)` should only be called after a POST / PUT request (ie. in a callback).
That way on further revalidation you get the updated data from the server anyway.
